modules/RTC/RTCUtils.js

```diff
@@ -1287,167 +1287,142 @@ class RTCUtils extends Listenable {
      * @returns {*} Promise object that will receive the new JitsiTracks
      */
     obtainAudioAndVideoPermissions(options = {}) {
-        const self = this;
-
         const dsOptions = {
             ...options.desktopSharingExtensionExternalInstallation,
             desktopSharingSources: options.desktopSharingSources
         };
 
-        return new Promise((resolve, reject) => {
-            const successCallback = function(stream) {
-                resolve(handleLocalStream(stream, options.resolution));
-            };
+        options.devices = options.devices || [ 'audio', 'video' ];
+        options.resolution = options.resolution || '720';
 
-            options.devices = options.devices || [ 'audio', 'video' ];
-            options.resolution = options.resolution || '720';
+        const requestingDesktop = options.devices.includes('desktop');
 
-            if (!screenObtainer.isSupported()
-                && options.devices.indexOf('desktop') !== -1) {
-                reject(new Error('Desktop sharing is not supported!'));
-            }
-            if (browser.isFirefox()
-
-                    // XXX The react-native-webrtc implementation that we
-                    // utilize on React Native at the time of this writing does
-                    // not support the MediaStream constructors defined by
-                    // https://www.w3.org/TR/mediacapture-streams/#constructors
-                    // and instead has a single constructor which expects (an
-                    // NSNumber as) a MediaStream ID.
-                    || browser.isReactNative()
-                    || browser.isTemasysPluginUsed()) {
-                const GUM = function(device, s, e) {
-                    this.getUserMediaWithConstraints(device, s, e, options);
-                };
+        if (requestingDesktop && !screenObtainer.isSupported()) {
+            return Promise.reject(
+                new Error('Desktop sharing is not supported!'));
+        }
+
+        let gumPromise;
+
+        if (browser.supportsMediaStreamConstructor()) {
```

Comment:
i think this name is a bit confusing.  if it wasn't for the comment on the `else` block below i wouldn't really get what was being checked for here.  i think at least the explaining comment should be here at the top, but wonder too if we could come up with a name that conveys what differentiates the two paths better.  maybe the `supportsMediaStreamConstructor` stays in the browser API, but we have another helper layer on top of that, something like "shouldGetAudioAndVideoStreamsSeparately()" which did this check inside of it?

Reply:
also, the if/else code is so long here...i wonder if it's worth moving the logic to functions to keep this one more readable.

Reply:
```
let gumPromise;
if (shouldGetAudioAndVideoStreamsSeparately()) {
  gumPromise = getAudioAndVideoStreamsSeparately()
} else {
  gumPromise = getAudioAndVideoStreams()
}
```
?